Feature: Allow user to dispatch list of benchmarks on a device #234

Kitsunp opened the pull request:
# Description

This PR implements support for dispatching multiple benchmarks to a single quantum device, addressing the need for comprehensive device characterization and comparison studies.

## Summary of changes:
- **Added `--all-benchmarks` flag** to run all available benchmarks (BSEQ, CLOPS, Quantum Volume, QML Kernel) on a specified device
- **Added `--except` flag** to exclude specific benchmarks when using `--all-benchmarks` (similar to pytest's exclude functionality)
- **Enhanced CLI parsing** in `metriq_gym/cli.py` to support new flags
- **Implemented dispatch logic** in `metriq_gym/run.py` with new functions:
  - `get_example_file_path()`: Maps benchmark types to example files
  - `dispatch_all_benchmarks()`: Orchestrates multiple benchmark execution
  - Enhanced `dispatch_job()` to handle both single and multi-benchmark modes
- **Updated README.md** with documentation for new multi-benchmark functionality
- **Added comprehensive tests** covering success cases, exclusions, error handling, and edge cases

## Dependencies:
No new dependencies required. The implementation uses existing infrastructure including:
- Example configuration files in `schemas/examples/`
- Current benchmark handlers and job management system
- Existing device setup and validation logic

# Issue ticket number and link

Fixes #234 - Feature: Allow user to dispatch list of benchmarks on a device

# Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

## Test Coverage
The implementation includes **10 comprehensive unit tests** in `tests/test_run.py`:

### Core functionality tests:
- `test_get_example_file_path()`: Verifies correct mapping of benchmark types to example files
- `test_dispatch_all_benchmarks_success()`: Tests successful execution of all benchmarks
- `test_dispatch_all_benchmarks_with_exclusions()`: Tests benchmark exclusion functionality
- `test_dispatch_all_benchmarks_missing_example_file()`: Tests graceful handling of missing files

### Integration tests:
- `test_dispatch_job_calls_all_benchmarks()`: Verifies dispatch_job routes to multi-benchmark flow
- `test_dispatch_job_calls_single_benchmark()`: Ensures backward compatibility with single benchmarks
- `test_dispatch_job_missing_input_file()`: Tests error handling for missing input files

### Existing tests (maintained):
- `test_setup_device_success()`: Device setup validation
- `test_setup_device_invalid_provider()`: Provider error handling
- `test_setup_device_invalid_device()`: Device error handling

## Running the tests:
```bash
# Run all tests
poetry run pytest tests/test_run.py -v

# Run only new multi-benchmark tests
poetry run pytest tests/test_run.py -k "dispatch_all_benchmarks or dispatch_job" -v

# Run with coverage
poetry run pytest tests/test_run.py --cov=metriq_gym.run --cov-report=term-missing
```

## Manual testing performed:
```bash
# Test all benchmarks
mgym dispatch --all-benchmarks --provider ibm --device ibm_sherbrooke

# Test with exclusions
mgym dispatch --all-benchmarks --provider ibm --device ibm_sherbrooke --except CLOPS "QML Kernel"

# Test backward compatibility
mgym dispatch schemas/examples/bseq.example.json --provider ibm --device ibm_sherbrooke
```

## Test configuration:
- **Python version**: 3.12+
- **Test framework**: pytest
- **Mocking**: unittest.mock with proper dataclass instances
- **Coverage**: All new functions and code paths covered
- **CI compatibility**: Tests use standard fixtures and follow existing patterns

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules (or not applicable)

Kitsunp commented:
@vprusso  Actually, it's kind of funny, but this concept could easily be extended to handle multiple devices, each with multiple benchmarks. Or, like the current setup here, where you can select a device and/or specific benchmarks, or run them one at a time. The issue, however, is that I lack the resources to actually test this out.

vprusso commented:
> @vprusso Actually, it's kind of funny, but this concept could easily be extended to handle multiple devices, each with multiple benchmarks. Or, like the current setup here, where you can select a device and/or specific benchmarks, or run them one at a time. The issue, however, is that I lack the resources to actually test this out.

Unfortunately access to run experiments on a variety of quantum devices *is indeed* somewhat cost prohibitive :)

Kitsunp commented:
A question in this repository, do all the contributors have to go through the pr? Or is it just for the event?

cosenal commented:
> A question in this repository, do all the contributors have to go through the pr? Or is it just for the event?

Yes, all contributions must go through a PR review process, including those from core maintainers of the project.

cosenal commented:
> > > A question in this repository, do all the contributors have to go through the pr? Or is it just for the event?
> > 
> > 
> > Yes, all contributions must go through a PR review process, including those from core maintainers of the project.
> 
> I guess the question is about whether all maintainers would go through the PR, and the answer is not always necessary but we want extra pairs of eyes for more suggestions?

ah, got it (@Changhao-Li sorry for the 👎 then :D)

Kitsunp commented:
Apparently, the issue was with the installation of the dependencies, but I don't recall modifying that.

cosenal commented:
> Apparently, the issue was with the installation of the dependencies, but I don't recall modifying that.

It looks unrelated to your change. I will look into it.

Kitsunp commented:
Thanks for the detailed feedback @cosenal and @Changhao-Li! You're absolutely right on all points.

## Issues I can resolve immediately:

1. **Multiple sources of truth**: I'll eliminate the hardcoded list in `run.py` and use `list(JobType)` directly from `metriq_gym/benchmarks/__init__.py` as the single source of truth.

2. **Error handling**: I'll improve error messages to include exception types and more context as suggested by @Changhao-Li.

3. **Test brittleness**: I'll make tests dynamic so they don't break when new benchmarks are added.

## Design discussion needed:

You're right that using example files as defaults is problematic. Before implementing a solution, I'd like to discuss the design approach:

**Option A**: Keep `--all-benchmarks` simplicity but with proper defaults defined in code (not example files)  
**Option B**: Multiple JSON files approach as you suggested: `mgym dispatch --configs custom_bseq.json custom_qv.json --provider ibm --device device`  

What's your preference? I want to make sure I implement what aligns with the project's vision for user experience.

I'll start with the technical fixes (sources of truth + error handling) while we discuss the configuration strategy.

cosenal commented:
> ## Design discussion needed:
> 
> You're right that using example files as defaults is problematic. Before implementing a solution, I'd like to discuss the design approach:
> 
> **Option A**: Keep `--all-benchmarks` simplicity but with proper defaults defined in code (not example files) **Option B**: Multiple JSON files approach as you suggested: `mgym dispatch --configs custom_bseq.json custom_qv.json --provider ibm --device device`
> 
> What's your preference? I want to make sure I implement what aligns with the project's vision for user experience.
> 
> I'll start with the technical fixes (sources of truth + error handling) while we discuss the configuration strategy.

OK, great that we are having a design discussion! I would say we go with Option B for the time being. Then in the future we can get fancier. We don't even need the `--configs` flag, we can assume that if the user passes multiple files, then all those benchmarks are executed. And in fact it could be also the same benchmark types multiple times with different parameters. What do you think? I don't foresee a case where one would want to run _all_ benchmark anyway. That'd be too costly, given that some of them may not scale well.

Kitsunp commented:
Perfect, thanks for the clarification @cosenal! 

Option B makes complete sense - it's much simpler and gives users full control. I agree that running *all* benchmarks would be too costly and not practical.

I'll implement the multiple JSON files approach without the `--configs` flag, so users can simply:
```bash
mgym dispatch file1.json file2.json file3.json --provider ibm --device device
```

This also nicely supports running the same benchmark type with different parameters as you mentioned. I'll get this implemented and remove all the `--all-benchmarks` functionality entirely.

Thanks for the guidance!

vprusso commented on metriq_gym/metriq_metadata.py:
```diff
@@ -23,7 +23,7 @@
     "ibm_oslo": 120,
     "ibm_perth": 22,
-    "ibm_prague": 56,   
+    "ibm_prague": 56,
```
Completely unrelated to this diff, but since there are edits here it reminded me that this file is here (and it likely should be removed). I would suggest we remove this file entirely. If @cosenal thinks this is outside of scope though, we can mark it as "to-be-removed" in a subsequent MR + issue. Either way.

Kitsunp commented:
**@cosenal** Thanks for the excellent feedback 

I've implemented the changes based on your suggestions:

## ✅ Changes Made:

### 1. **Simplified exception handling** (lines ~150-162):
- **Before**: 13 lines with complex traceback logic + DRY violation (repeated message 3x)
- **After**: 3 clean lines: `error_details = f"{type(e).__name__}: {str(e)}"` 
- **Result**: Maintains useful error info without unnecessary complexity

### 2. **Added device to dispatch message**:
- **Before**: `"Dispatching BSEQ benchmark from config.json..."`
- **After**: `"Dispatching BSEQ benchmark from config.json on ibm_sherbrooke..."`
- Updated README to reflect actual output

### 3. **Simplified docstring** following **Rotter's DRY principle**:
- **Before**: 29 lines that literally duplicated the code logic step-by-step
- **After**: 8 lines focusing on **business purpose** and **design decisions**
- **Removed**: Workflow section that violated DRY by repeating implementation

## 📚 Applied Rotter's principle: 
> "A comment should not just repeat what the code is already saying" 

The code is now self-explanatory, docstring explains **why** not **what**.

The LLM comment was spot-on 😄 - detailed docs are valuable when they add context the code can't express, but not when they duplicate the implementation!

**Ready for review** ✨

Kitsunp commented:
**@cosenal** Thanks for the additional feedback! 

Applied both suggestions:

## ✅ Changes Made:

### 1. **Optimized `get_available_benchmarks()`**:
- **Before**: Returns `list[JobType]` → required `[jt.value for jt in available_benchmarks]`
- **After**: Returns `list[str]` directly → used without conversion
- **Result**: Cleaner code in `dispatch_job()`, no unnecessary type conversion

### 2. **Removed trivial test**:
- **Removed**: `test_get_available_benchmarks_dynamic()`
- **Reason**: As you noted, it was testing basic enum functionality and content that could easily get outdated
- **Kept**: All meaningful functional tests

Kitsunp commented:
@cosenal Quick question about the unitaryHACK workflow - I have another PR where I'm already assigned to the mirror_circuit issue, but the platform hasn't detected/tracked it yet for the event. 

Will both PRs be merged normally, or will you close the issue to mark completion for the hackathon while keeping the PRs open? I want to make sure both contributions get properly tracked for unitaryHACK 2025 before the June 11th deadline.

Thanks for clarifying the process! 🚀
